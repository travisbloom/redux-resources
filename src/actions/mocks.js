import {
    mockStore,
    normalizer,
    RESOURCE_REDUCER,
    RESOURCE,
    RESOURCE_ID,
    INITIAL_STATE,
    INITIAL_STATE_WITH_CACHED_RESOURCE,
    QUERY,
    RESOURCE_UPDATE_PAYLOAD,
    formatErrors,
} from '../mocks'
import createResource from './createResource'
import deleteResource from './deleteResource'
import listResource from './listResource'
import partialUpdateResource from './partialUpdateResource'
import updateResource from './updateResource'
import retrieveResource from './retrieveResource'

const generateCreateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState)
    const { createMockResources } = createResource({
        resource: RESOURCE_REDUCER,
        normalizer,
        request,
        formatErrors,
    })
    await store.dispatch(createMockResources(RESOURCE, options))
    return store.getActions()
}

const generateDeleteResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { deleteMockResources } = deleteResource({
        resource: RESOURCE_REDUCER,
        request,
        formatErrors,
    })
    await store.dispatch(deleteMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generateListResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { listMockResources } = listResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(listMockResources(QUERY, options))
    return store.getActions()
}

const generateRetrieveResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { retrieveMockResources } = retrieveResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(retrieveMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generatePartialUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { partialUpdateMockResources } = partialUpdateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(partialUpdateMockResources(RESOURCE_ID, RESOURCE_UPDATE_PAYLOAD, options))
    return store.getActions()
}

const generateUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { updateMockResources } = updateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    const body = { ...RESOURCE, ...RESOURCE_UPDATE_PAYLOAD }
    await store.dispatch(updateMockResources(RESOURCE_ID, body, options))
    return store.getActions()
}

module.exports = {
    generateCreateResourceActions,
    generateDeleteResourceActions,
    generateListResourceActions,
    generateRetrieveResourceActions,
    generatePartialUpdateResourceActions,
    generateUpdateResourceActions,
}
