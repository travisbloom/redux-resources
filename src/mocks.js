import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Schema, normalize, arrayOf } from 'normalizr'

import { getResourceInitialState } from './reducers'

const RESOURCE_REDUCER = 'mockResources'
const RELATED_RESOURCE_REDUCER = 'relatedMockResources'

const RELATED_RESOURCE_ID = 'relatedResourceId'
const RELATED_RESOURCE = {
    id: RELATED_RESOURCE_ID,
    relatedProp1: 'initial relatedProp1',
}
const RESOURCE_ID = 'resourceId'
const RESOURCE = {
    id: RESOURCE_ID,
    prop1: 'initial prop1',
    prop2: 'initial prop2',
    relatedResource: RELATED_RESOURCE,
}
const NORMALIZED_RESOURCE = {
    id: RESOURCE_ID,
    prop1: 'initial prop1',
    prop2: 'initial prop2',
    relatedResource: RELATED_RESOURCE_ID,
}

const RESOURCE_UPDATE_PAYLOAD = {
    id: RESOURCE_ID,
    prop2: 'UPDATED PROP_1',
}

const SERVER_ERRORS = [
    'This is an initial error',
    'This is another error',
]

const QUERY = {
    limit: 30,
    offset: 0,
    someQueryParam: 'someValue',
}
const QUERY_STRING = 'key=30&key=0&key="someValue"'

const INITIAL_STATE = {
    resources: {
        [RELATED_RESOURCE_REDUCER]: getResourceInitialState(),
        [RESOURCE_REDUCER]: getResourceInitialState(),
    },
}
const INITIAL_STATE_WITH_LIST_BEING_FETCHED = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resourceListsBeingFetched: [QUERY_STRING],
        },
    },
}
const INITIAL_STATE_WITH_CACHED_LIST = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resourceLists: {
                [QUERY_STRING]: {
                    result: [RESOURCE_ID],
                },
            },
        },
    },
}
const INITIAL_STATE_WITH_CACHED_AND_SELECTED_LIST = {
    resources: {
        ...INITIAL_STATE_WITH_CACHED_LIST.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE_WITH_CACHED_LIST.resources[RESOURCE_REDUCER],
            selectedResourceList: QUERY_STRING,
        },
    },
}
const INITIAL_STATE_WITH_RESOURCE_BEING_FETCHED = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resourcesBeingFetched: [RESOURCE_ID],
        },
    },
}
const INITIAL_STATE_WITH_CACHED_RESOURCE = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resources: {
                [RESOURCE_ID]: NORMALIZED_RESOURCE,
            },
        },
    },
}
const INITIAL_STATE_WITH_CACHED_AND_SELECTED_RESOURCE = {
    resources: {
        ...INITIAL_STATE_WITH_CACHED_RESOURCE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE_WITH_CACHED_RESOURCE.resources[RESOURCE_REDUCER],
            selectedResource: RESOURCE_ID,
        },
    },
}
const INITIAL_STATE_WITH_CACHED_RESOURCE_AND_CACHED_RESOURCE_LIST = {
    resources: {
        ...INITIAL_STATE.resources,
        [RESOURCE_REDUCER]: {
            ...INITIAL_STATE.resources[RESOURCE_REDUCER],
            resources: {
                [RESOURCE_ID]: NORMALIZED_RESOURCE,
            },
            resourceLists: {
                [QUERY_STRING]: {
                    result: [RESOURCE_ID],
                },
            },
        },
    },
}

const resourceSchema = new Schema(RESOURCE_REDUCER)
const relatedResourceSchema = new Schema(RELATED_RESOURCE_REDUCER)
resourceSchema.define({
    relatedResource: relatedResourceSchema,
})
const normalizer = (resources) => {
    const array = Array.isArray(resources) ? resources : [resources]
    return normalize(array, arrayOf(resourceSchema))
}
const formatErrors = errors => errors

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

module.exports = {
    RESOURCE_REDUCER,
    RESOURCE_ID,
    RESOURCE,

    INITIAL_STATE,
    INITIAL_STATE_WITH_LIST_BEING_FETCHED,
    INITIAL_STATE_WITH_CACHED_LIST,
    INITIAL_STATE_WITH_CACHED_AND_SELECTED_LIST,
    INITIAL_STATE_WITH_RESOURCE_BEING_FETCHED,
    INITIAL_STATE_WITH_CACHED_RESOURCE,
    INITIAL_STATE_WITH_CACHED_AND_SELECTED_RESOURCE,
    INITIAL_STATE_WITH_CACHED_RESOURCE_AND_CACHED_RESOURCE_LIST,

    RESOURCE_UPDATE_PAYLOAD,

    SERVER_ERRORS,
    QUERY,
    QUERY_STRING,

    mockStore,
    formatErrors,
    normalizer,
}
