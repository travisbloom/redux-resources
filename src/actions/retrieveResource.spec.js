import {
    RESOURCE,
    SERVER_ERRORS,
    INITIAL_STATE_WITH_CACHED_RESOURCE,
    INITIAL_STATE_WITH_RESOURCE_BEING_FETCHED,
    INITIAL_STATE_WITH_CACHED_AND_SELECTED_RESOURCE,
} from '../mocks'
import { generateRetrieveResourceActions } from './mocks'

const request = () => Promise.resolve(RESOURCE)
const errorRequest = () => Promise.reject(SERVER_ERRORS)

it('will dispatch two actions on success', async () => {
    const actions = await generateRetrieveResourceActions({ request })
    expect(actions.length).toBe(2)
})

it('will dispatch two actions on error', async () => {
    const actions = await generateRetrieveResourceActions({ request: errorRequest })
    expect(actions.length).toBe(2)
})

it('will dispatch one action if the resource is cached but not selected', async () => {
    const actions = await generateRetrieveResourceActions({
        request: errorRequest,
        initialState: INITIAL_STATE_WITH_CACHED_RESOURCE,
    })
    expect(actions.length).toBe(1)
})

it('will dispatch two action if the resource is cached but shouldIgnoreCache is passed', async () => {
    const actions = await generateRetrieveResourceActions({
        request: errorRequest,
        shouldIgnoreCache: true,
        initialState: INITIAL_STATE_WITH_CACHED_RESOURCE,
    })
    expect(actions.length).toBe(2)
})

it('will dispatch no actions if the resource is cached and selected', async () => {
    const actions = await generateRetrieveResourceActions({
        request: errorRequest,
        initialState: INITIAL_STATE_WITH_CACHED_AND_SELECTED_RESOURCE,
    })
    expect(actions.length).toBe(0)
})

it('will dispatch no actions if the resource is being fetched', async () => {
    const actions = await generateRetrieveResourceActions({
        request: errorRequest,
        initialState: INITIAL_STATE_WITH_RESOURCE_BEING_FETCHED,
    })
    expect(actions.length).toBe(0)
})

it('will have a update selectedResource action', async () => {
    const [inititalAction] = await generateRetrieveResourceActions({
        request: errorRequest,
        initialState: INITIAL_STATE_WITH_CACHED_RESOURCE,
    })
    expect(inititalAction).toMatchSnapshot()
})

it('will have an initial action', async () => {
    const [inititalAction] = await generateRetrieveResourceActions({ request })
    expect(inititalAction).toMatchSnapshot()
})

it('will have a success action', async () => {
    const [_, successAction] = await generateRetrieveResourceActions({ request })
    expect(successAction).toMatchSnapshot()
})

it('will have a error action', async () => {
    const [_, errorAction] = await generateRetrieveResourceActions({ request: errorRequest })
    expect(errorAction).toMatchSnapshot()
})
