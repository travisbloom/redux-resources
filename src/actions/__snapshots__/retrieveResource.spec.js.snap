exports[`test will have a error action 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "isSelected": true,
    "reduxResourcesActionType": "RETRIEVE_RESOURCE_REQUEST_ERROR",
    "resource": "mockResources",
    "timestamp": "2016-01-01T05:00:00.000Z",
  },
  "payload": Array [
    "This is an initial error",
    "This is another error",
  ],
  "type": "GET_MOCK_RESOURCES_REQUEST_ERROR",
}
`;

exports[`test will have a success action 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "isSelected": true,
    "normalizedResponse": Object {
      "entities": Object {
        "mockResources": Object {
          "resourceId": Object {
            "id": "resourceId",
            "prop1": "initial prop1",
            "prop2": "initial prop2",
            "relatedResource": "relatedResourceId",
          },
        },
        "relatedMockResources": Object {
          "relatedResourceId": Object {
            "id": "relatedResourceId",
            "relatedProp1": "initial relatedProp1",
          },
        },
      },
      "result": Array [
        "resourceId",
      ],
    },
    "reduxResourcesActionType": "RETRIEVE_RESOURCE_REQUEST_SUCCESS",
    "resource": "mockResources",
    "timestamp": "2016-01-01T05:00:00.000Z",
  },
  "payload": Object {
    "id": "resourceId",
    "prop1": "initial prop1",
    "prop2": "initial prop2",
    "relatedResource": Object {
      "id": "relatedResourceId",
      "relatedProp1": "initial relatedProp1",
    },
  },
  "type": "GET_MOCK_RESOURCES_REQUEST_SUCCESS",
}
`;

exports[`test will have a update selectedResource action 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "isSelected": true,
    "reduxResourcesActionType": "SELECT_RESOURCE",
    "resource": "mockResources",
    "timestamp": "2016-01-01T05:00:00.000Z",
  },
  "type": "SELECT_MOCK_RESOURCES",
}
`;

exports[`test will have an initial action 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "isSelected": true,
    "reduxResourcesActionType": "RETRIEVE_RESOURCE_REQUEST",
    "resource": "mockResources",
    "timestamp": "2016-01-01T05:00:00.000Z",
  },
  "payload": "resourceId",
  "type": "GET_MOCK_RESOURCES_REQUEST",
}
`;
